i6_models/primitives/feature_extraction.py

```diff
@@ -111,3 +117,80 @@ def forward(self, raw_audio, length) -> Tuple[torch.Tensor, torch.Tensor]:
             length = ((length - self.n_fft) // self.hop_length) + 1
 
         return feature_data, length.int()
+
+
+@dataclass
+class RasrCompatibleLogMelFeatureExtractionV1Config(ModelConfiguration):
+    """
+    Attributes:
+        sample_rate: audio sample rate in Hz
+        win_size: window size in seconds
+        hop_size: window shift in seconds
+        min_amp: minimum amplitude for safe log
+        num_filters: number of mel windows
+    """
+
+    sample_rate: int
+    win_size: float
+    hop_size: float
+    min_amp: float
+    num_filters: int
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        assert self.win_size > 0 and self.hop_size > 0, "window settings need to be positive"
+        assert self.num_filters > 0, "number of filters needs to be positive"
+        assert self.hop_size <= self.win_size, "using a larger hop size than window size does not make sense"
+
+
+class RasrCompatibleLogMelFeatureExtractionV1(nn.Module):
+    """
+    Rasr-compatible log-mel feature extraction using log10. Does not use torchaudio.
+
+    Using it wrapped with torch.no_grad() is recommended if no gradient is needed
+    """
+
+    def __init__(self, cfg: RasrCompatibleLogMelFeatureExtractionV1Config):
+        super().__init__()
+        self.hop_length = int(cfg.hop_size * cfg.sample_rate)
+        self.min_amp = cfg.min_amp
+        self.win_length = int(cfg.win_size * cfg.sample_rate)
+        # smallest power if two which is greater than or equal to win_length
+        self.n_fft = 2 ** math.ceil(math.log2(self.win_length))
+
+        self.register_buffer(
+            "mel_basis",
+            torch.tensor(
+                filters.mel(
+                    sr=cfg.sample_rate,
+                    n_fft=self.n_fft,
+                    n_mels=cfg.num_filters,
+                    fmin=0,
+                    fmax=cfg.sample_rate // 2,
+                    htk=True,
+                    norm=None,
+                ),
+            ),
+        )
+        self.register_buffer("window", torch.hann_window(self.win_length, periodic=False))
+
+    def forward(self, raw_audio, length) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        :param raw_audio: [B, T]
+        :param length in samples: [B]
+        :return features as [B,T,F] and length in frames [B]
+        """
+        windowed = raw_audio.unfold(1, size=self.win_length, step=self.hop_length)  # [B, T', W=win_length]
+        smoothed = windowed * self.window.unsqueeze(0)  # [B, T', W]
```

Comment:
There is preemphasis missing compared to the default rasr feature extraction. should we add it?

Reply:
Rasr again [rasr](https://pytorch.org/audio/stable/generated/torchaudio.functional.preemphasis.html) does this differently than for example [torchaudio](https://pytorch.org/audio/stable/generated/torchaudio.functional.preemphasis.html), as Rasr keeps the last state from last segment(?).

Reply:
> Rasr keeps the last state from last segment(?).

So RASR is stateful? This function here is stateless, and I think that makes more sense. I would not add this. Not even sure how you would add this, as you don't know the last segment here in any case.

Reply:
I guess this is only a problem in streaming. But this feature extraction is mostly for training where we see the full utterance and thus don't need to worry about state-carryover. So it would be nice to add it (as an option).

Reply:
maybe we should change the name of the class, maybe to `RasrReproducibleLogMelFeatureExtractionV1Config`? @michelwi @curufinwe

Reply:
> > I guess this is only a problem in streaming. But this feature extraction is mostly for training where we see the full utterance and thus don't need to worry about state-carryover. So it would be nice to add it (as an option).
> 
> but if we don't do preemphais in training, we would also not do in recognition in RASR for the compatibility.

Correct.

> maybe we should change the name of the class, maybe to `RasrReproducibleLogMelFeatureExtractionV1Config`? @michelwi @curufinwe

If you want, I have no preference.

Reply:
FYI: Adding preemphasis improved the WER from 15.9 to 15.6 one some random model training I did to test the pipeline.